src/vm-memory/src/lib.rs

```diff
@@ -5,18 +5,355 @@
+use std::io::Error as IoError;
+use std::os::unix::io::AsRawFd;
 
-// Re-export only what is needed in Firecracker.
-pub use vm_memory_upstream::{
-    address, Address, ByteValued, Bytes, Error, FileOffset, GuestAddress, GuestMemory,
-    GuestMemoryError, GuestMemoryRegion, MemoryRegionAddress, MmapRegion,
+use vm_memory_upstream::bitmap::AtomicBitmap;
+use vm_memory_upstream::mmap::{check_file_offset, NewBitmap};
+use vm_memory_upstream::{
+    GuestMemoryMmap as UpstreamGuestMemoryMmap, GuestRegionMmap as UpstreamGuestRegionMmap,
+    MmapRegion as UpstreamMmapRegion,
 };
+
+pub type GuestMemoryMmap = UpstreamGuestMemoryMmap<Option<AtomicBitmap>>;
+pub type GuestRegionMmap = UpstreamGuestRegionMmap<Option<AtomicBitmap>>;
+pub type MmapRegion = UpstreamMmapRegion<Option<AtomicBitmap>>;
+
+const GUARD_NUMBER: usize = 2;
+
+/// Build a MmapRegion surrounded by guard pages.
+fn build_guarded_region(
+    file_offset: Option<FileOffset>,
+    size: usize,
+    prot: i32,
+    flags: i32,
+    track_dirty_pages: bool,
+) -> Result<MmapRegion, MmapRegionError> {
+    let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+    // Create the guarded range size (received size + X pages),
+    // where X is defined as a constant GUARD_NUMBER.
+    let guarded_size = size + GUARD_NUMBER * page_size;
+
+    // Map the guarded range to PROT_NONE
+    let guard_addr = unsafe {
+        libc::mmap(
+            std::ptr::null_mut(),
+            guarded_size,
+            libc::PROT_NONE,
+            libc::MAP_ANONYMOUS | libc::MAP_PRIVATE | libc::MAP_NORESERVE,
+            -1,
+            0,
+        )
+    };
+
+    if guard_addr == libc::MAP_FAILED {
+        return Err(MmapRegionError::Mmap(IoError::last_os_error()));
+    }
+
+    let (fd, offset) = if let Some(ref f_off) = file_offset {
+        check_file_offset(f_off, size)?;
+        (f_off.file().as_raw_fd(), f_off.start())
+    } else {
+        (-1, 0)
+    };
+
+    let map_addr = guard_addr as usize + page_size * (GUARD_NUMBER / 2);
+
+    // Inside the protected range, starting with guard_addr + PAGE_SIZE,
+    // map the requested range with received protection and flags
+    let addr = unsafe {
+        libc::mmap(
+            map_addr as *mut libc::c_void,
+            size,
+            prot,
+            flags | libc::MAP_FIXED,
+            fd,
+            offset as libc::off_t,
+        )
+    };
+
+    if addr == libc::MAP_FAILED {
+        return Err(MmapRegionError::Mmap(IoError::last_os_error()));
+    }
+
+    let bitmap = match track_dirty_pages {
+        true => Some(AtomicBitmap::with_len(size)),
+        false => None,
+    };
+
+    unsafe {
+        MmapRegionBuilder::new_with_bitmap(size, bitmap)
+            .with_raw_mmap_pointer(addr as *mut u8)
+            .with_mmap_prot(prot)
+            .with_mmap_flags(flags)
+            .build()
+    }
+}
+
+/// Helper for creating the guest memory.
+pub fn create_guest_memory(
```

Comment:
This vm-memory crate was purposed as a proxy crate (to allow specialization or adding functionality missing in upstream atm) until the feature gap closes. Dirty page tracking in upstream brings use one step closer to dropping this crate from FC.

This PR this is challenging that by further expanding on functionality/scope.  Do you think we should revisit that decision or strip it as much as possible in this PR ?

Reply:
> This PR this is challenging that by further expanding on functionality/scope. Do you think we should revisit that decision or strip it as much as possible in this PR ?

It is not expanding any functionality besides what already existed in Firecracker main branch.
The guarded pages have already been merged and used for some time. This PR is just removing them in a commit and porting them to the new code in a follow-up commit